### ISSUE TITLE
NullReferenceException when using AsyncPool

### ISSUE THREAD

**Issue by wassim-k** [microsoft/FASTER#784]
I'm using [FasterWrapper](https://github.com/microsoft/FASTER/blob/main/cs/playground/AsyncStress/FasterWrapper.cs) as suggested by https://github.com/microsoft/FASTER/issues/780#issuecomment-1383267812
But when running the following code, which checkpoints from multiple threads, I'm receving a `NullReferenceException` at `var status = session.Upsert(key, value);`

Can you please explain why?

```csharp
var faster = new FasterWrapper<long, long>(isRefType: false, useLargeLog: false, useOsReadBuffering: false);
await Task.WhenAll(Enumerable.Range(0, 2).Select(async i =>
{
    faster.Upsert(i, i);
    await faster.Store.TakeHybridLogCheckpointAsync(CheckpointType.FoldOver);
}));
```

I've also noticed in the synchronous variation of methods like `Upsert` in `FasterWrapper`, `_sessionPool.GetAsync().GetAwaiter().GetResult()` is being used instead of the synchronous `_sessionPool.Get()`, is there reason for that?

**Comment by tcsaddul** [comment]
Maybe the same problem here when I upgraded to the latest version that error appears in the Upsert in concurrent calls. After reverting to the previous version, error disappeared.

**Comment by wassim-k** [comment]
Worth mentioning I've had this issue happen in 2.2.0 & 2.3.0 versions.

**Comment by badrishc** [comment]
Can you summarize in detail what's the exact issue in the latest version. Where is the exception, what's the call stack, have your tried any fixes yourself etc.

For sync, you can check the code to see if there is a good reason. I think it's because Try might fail.

**Comment by wassim-k** [comment]
Reproduction:
* Checkout AsyncStress playground
* Expose `public FasterKV<Key, Value> Store => _store;` from `FasterWrapper<Key, Value>`
* Replace Program's `Main` method with above code as is.

Exception:
```
Unhandled exception. System.NullReferenceException: Object reference not set to an instance of an object.
   at FASTER.core.ClientSession`6.InternalFasterSession.TryLockEphemeralExclusive(RecordInfo& recordInfo)
   at AsyncStress.FasterWrapper`2.Upsert(Key key, Value value) in C:\Code\AsyncStress\FasterWrapper.cs:line 87
   at AsyncStress.Program.<>c__DisplayClass10_0.<<Main>b__0>d.MoveNext() in C:\Code\AsyncStress\Program.cs:line 70
--- End of stack trace from previous location ---
   at AsyncStress.Program.Main(String[] args) in C:\Code\AsyncStress\Program.cs:line 68
   at AsyncStress.Program.<Main>(String[] args)
```

The issue happens in **v2.2.0** and **v2.3.0** but not in **v.2.1.0**

**Comment by badrishc** [comment]
Fixed in https://github.com/microsoft/FASTER/pull/785
